jlm/llvm/opt/alias-analyses/PointsToGraphAliasAnalysis.cpp

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2025 Håvard Krogstie <krogstie.havard@gmail.com>
+ * See COPYING for terms of redistribution.
+ */
+
+#include <jlm/llvm/opt/alias-analyses/PointsToGraphAliasAnalysis.hpp>
+
+namespace jlm::llvm::aa
+{
+PointsToGraphAliasAnalysis::PointsToGraphAliasAnalysis(PointsToGraph & pointsToGraph)
+    : PointsToGraph_(pointsToGraph)
+{}
+
+PointsToGraphAliasAnalysis::~PointsToGraphAliasAnalysis() = default;
+
+std::string
+PointsToGraphAliasAnalysis::ToString() const
+{
+  return "PointsToGraphAA";
+}
+
+AliasAnalysis::AliasQueryResponse
+PointsToGraphAliasAnalysis::Query(
+    const rvsdg::Output & p1,
+    const size_t s1,
+    const rvsdg::Output & p2,
+    const size_t s2)
+{
+  // If the two pointers are the same value, they must alias
+  if (&p1 == &p2)
+    return MustAlias;
+
+  // Assume that all pointers actually exist in the PointsToGraph
+  auto & p1RegisterNode = PointsToGraph_.GetRegisterNode(p1);
+  auto & p2RegisterNode = PointsToGraph_.GetRegisterNode(p2);
+
+  // Check if both pointers may target the external node, to avoid iterating over large sets
+  const auto & externalNode = PointsToGraph_.GetExternalMemoryNode();
+  if (p1RegisterNode.HasTarget(externalNode) && p2RegisterNode.HasTarget(externalNode))
+    return MayAlias;
+
+  // Quickly checks if the given register node has only one possible target
+  const auto GetSingleTarget = [&](const PointsToGraph::RegisterNode & node,
+                                   size_t size) -> std::optional<const PointsToGraph::MemoryNode *>
+  {
+    std::optional<const PointsToGraph::MemoryNode *> singleTarget;
+    for (auto & target : node.Targets())
+    {
+      // Skip memory locations that are too small to hold size
+      const auto targetSize = GetMemoryNodeSize(target);
+      if (targetSize.has_value() && *targetSize < size)
+        continue;
+
+      if (singleTarget.has_value())
+        return std::nullopt;
+
+      singleTarget = &target;
+    }
+
+    return singleTarget;
+  };
```

Comment:
You can make this lambda a private function to improve readability in this method.